engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArray.java

```diff
@@ -1577,6 +1581,13 @@ public String toString() {
     @Override
     public boolean equals(Object obj) {
         if (obj instanceof NDArray) {
+            // do no compare content if obj is released
+            if (((NDArray) obj).isReleased()) {
+                return this == obj;
+            }
+            if (((NDArray) obj).getManager() != manager) {
```

Comment:
`equals()` compares the content of two NDArray, if their NDManager are different, their content still can be equals

Reply:
NDScope.unregister calls resources.remove which uses equals to identify the object. 
I got an exception from unregister in two cases: one object (not the one to be unregistered) in NDScope was on a different Device and one was already closed. The lines I added - maybe not perfect - were intended to solve both cases.

Reply:
I see, we should not use `.remove()`, we can call `.pop()` instead

Reply:
> I see, we should remove the NDArray by it's uid

I think this is not save because the uid from the handle is only unique as long as the handle lives. If you create a uuid by uid plus handleCreationTimestamp it would do.

Reply:
You are right, uid meant to be unique, but it's possible have a clash if the NDArray is closed and the same memory get reused. We should add a timestamp to it.

Reply:
Do you make the corrections to the equals method?

Reply:
I should not change `equals()` definition. `equals()` only mean the value are the same. Not the same instance. In NDScope, we must ensure we are deleting the same object. If two NDArray with the same value pushed into NDScope, we might deleted the wrong one.

Reply:
It is an option to solve in NDScope. However, the problems I experience where Exceptions coming from the equals-Method. That is not part of the equals method contract - it should answer yes or no.
If you would like to keep equals as it is. A simple Wrapper of NDArray would do that is only used inside NDScope to put into the list. There you could without an equals implementation just inherit the one from Object. If you go with uid and timestamp equals than the list should be reset by a Set as in a list could exist many NDArrays that are "equal" according to content but the remove method would only remove the first occurance.

Reply:
Let separate in different issues:
1. NDScope should track the object itself, the ArrayList might not be the right collection, we can use `IdentityHashMap` to resolve this problem
2. uid is not unique. we can solve this problem in a different PR
3. `equals()` function: `isReleased()` is a bug, we can fixed it or in a different PR

Reply:
I comitted a suggestion, but I think it is not correct ... NDArrayWrapper and NDArrayList it a problem ... I will fix it.

Reply:
Thx. IdentityHashMap works nicely. Better solution than my wrapper suggestion ... and using NDArrayList with the wrapper is simply wrong.

Reply:
1. done?
Points 2. and 3. in a separate PR?